activities.el

```diff
@@ -801,6 +819,90 @@ activity's name is NAME."
 
+(defvar activities--age-spec
+  `((?Y "year"   "years"   ,(round (* 60 60 24 365.2425)))
+    (?M "month"  "months"  ,(round (* 60 60 24 30.436875)))
+    (?w "week"   "weeks"   ,(* 60 60 24 7))
+    (?d "day"    "days"    ,(* 60 60 24))
+    (?h "hour"   "hours"   ,(* 60 60))
+    (?m "minute" "minutes" 60)
+    (?s "second" "seconds" 1))
+  "Age specification.  See `magit--age-spec', which this duplicates.")
+
+(defun activities--age (age &optional abbreviate)
+  "Summarize AGE and possibly ABBREVIATE.
+Adapted from `magit--age'."
+  (cl-labels ((fn (age spec)
+                (pcase-let ((`(,char ,unit ,units ,weight) (car spec)))
+                  (let ((cnt (round (/ age weight 1.0))))
+                    (if (or (not (cdr spec))
+                            (>= (/ age weight) 1))
+                        (list cnt (cond (abbreviate char)
+                                        ((= cnt 1) unit)
+                                        (t units)))
+                      (fn age (cdr spec)))))))
+    (fn age activities--age-spec)))
+
+(defun activities--oldest-age (activities)
+  "Return the age in seconds of the oldest activity in ACTIVITIES."
+  (cl-loop for (_name . act) in activities maximize
+	   (cl-loop
+	    for type in '(default last)
+	    for func = (intern (format "activities-activity-%s" type))
+	    maximize
+	    (float-time
+	     (time-since
+	      (map-elt (activities-activity-state-etc (funcall func act)) 'time))))))
+
+(defun activities-annotate (max-age oldest-possible)
+  "Return an activity annotation function.
+MAX-AGE is the maximum age of any activity in seconds.
+OLDEST-POSSIBLE is the oldest age in the `vc-annotate-color-map'."
+  (lambda (name)
+    (when-let ((activity (map-elt activities-activities name)))
+      (let (data (age-len 14))
+	(dolist (type '(last default))
+	  (let* ((func (intern (format "activities-activity-%s" type)))
+		 (state (funcall func activity))
+		 (time (map-elt (activities-activity-state-etc state) 'time))
+		 (window-state (activities-activity-state-window-state state))
+		 (buffers (window-state-buffers window-state))
+		 (files ( activities--map-window-state-leafs window-state
+			  (lambda (l)
+			    (bookmark-get-filename
+			     (activities-buffer-bookmark
+			      (map-nested-elt (cdr l)
+					      '(parameters activities-buffer))))))))
+	    (setf (alist-get type data)
+		  (list (length buffers)
+			(length (delq nil files))
+			(float-time (time-since time))))))
+	(pcase-let* ((`(,lastb ,lastf ,lastt) (map-elt data 'last))
+		     (`(,defb ,deff ,deft) (map-elt data 'default))
+		     (age (min lastt deft))
+		     (ann (format "%s:%s|%s %s:%s|%s "
+				  (propertize "buffers" 'face 'bold)
+				  (propertize (format "%2d" lastb) 'face 'success)
+				  (propertize (format "%-2d" defb) 'face 'warning)
+				  (propertize "files" 'face 'bold)
+				  (propertize (format "%2d" lastf) 'face 'success)
+				  (propertize (format "%-2d" deff) 'face 'warning)))
```

Comment:
This looks nice, but the variable names are a bit unclear with those abbreviations.  I'd rather the code be clear and unambiguous, even if it means the formatting suffers.  So please just use the full words in the variable names.